ow_power_system/src/PowerSystemPack.cpp

```diff
@@ -0,0 +1,784 @@
+  // the bus each loop to trigger predictions.
+  // This loop should run at 0.5Hz, publishing predictions every 2 seconds via rostopics.
+  int startLoops = 2 * NUM_NODES;
+  while(ros::ok())
+  {
+    ros::spinOnce();
+
+    // Set up value modifiers in each node for injection later.
+    injectFaults();
+
+    // DEBUG PRINT
+    if (m_timestamp_print_debug)
+    {
+      ROS_INFO_STREAM("Timestamp " << m_nodes[0].model.timestamp);
+    }
+
+    // Warning message for excessive power draw.
+    std::string highDrawMsg = "";
+    int msgIndex = 0;
+    bool firstLine = true;
+
+    // Cycle the power stats in the nodes.
+    for (int i = 0; i < NUM_NODES; i++)
+    {
+      double excessiveDraw = m_nodes[i].node.RunOnce();
+      // Add to the warning message if the draw exceeded set limits (which returns
+      // a value greater than 0).
+      if (excessiveDraw > 0)
+      {
+        // Formatting for the warning message, to display up to 4 nodes
+        // per line for conciseness.
+        std::string header = "N";
+        if (i < 10)
+        {
+          header += "0";
+        }
+        header += std::to_string(i) + ": " + std::to_string(excessiveDraw) + "W";
+        switch(msgIndex)
+        {
+          case 0:
+            if (firstLine)
+            {
+              firstLine = false;
+            }
+            else
+            {
+              highDrawMsg += ",\n";
+            }
+            highDrawMsg += header;
+            msgIndex++;
+            break;
+          case 1:
+            highDrawMsg += ", " + header;
+            msgIndex++;
+            break;
+          case 2:
+            highDrawMsg += ", " + header;
+            msgIndex++;
+            break;
+          case 3:
+            highDrawMsg += ", " + header;
+            msgIndex = 0;
+            break;
+          default:
+            break;
+        }
+      }
+
+      m_nodes[i].node.GetPowerStats(m_nodes[i].model.timestamp, m_nodes[i].model.wattage,
+                                    m_nodes[i].model.voltage, m_nodes[i].model.temperature);
+      
+      // If the timestamp is the same as the previous one (happens during startup),
+      // do not publish the data to prevent crashes.
+      if (m_nodes[i].previous_time != m_nodes[i].model.timestamp)
+      {
+        auto timestamp = START_TIME + std::chrono::milliseconds(
+          static_cast<unsigned>(m_nodes[i].model.timestamp * 1000)
+        );
+        
+        // Compile a vector<shared_ptr<DoubleMessage>> and then individually
+        // publish each component.
+        std::vector<std::shared_ptr<DoubleMessage>> data_to_publish;
+        double input_power, input_tmp, input_voltage;
+
+        // HACK ALERT (May 2023):
+        // The first batch of data (sent in during startup) seems to be lost to 
+        // the void. This is a bandaid fix to send the initial data in twice
+        // for each node on the next loop.
```

Comment:
GSAP treats the initial data sent into its prognosers as the starting conditions of the battery, which is a different and specific state than the randomly distributed wattage/voltage/temperature generated during main cycles. If you don't send in this initial data and instead jump straight to the generated values, the starting RUL/SoC values are invalid (RUL starts anywhere from 1-3k and SoC starts at 7-15%).

I found previously that unless this initial data was sent in twice, GSAP seemingly wouldn't receive it and the prognosers would output invalid predictions. However, I did some extensive testing this morning and it I don't observe the problem behavior anymore with sending in the starting data once to each prognoser. I'm not sure what changed to address the issue (it may very well have been an issue with my own past implementation), but I'll go ahead and remove this stipulation since I can't replicate this issue anymore.